pkg/testing/helm/helm_test.go

```diff
@@ -126,6 +128,108 @@ func (h *HelmTestSuite) TestComponentsFromManifestAfterDryRunHelmInstall(c *C) {
+	nodeSelector := map[string]string{
+		"selector-key": "selector-value",
+	}
+	toleration := []corev1.Toleration{
+		{
+			Key:      "taint-key",
+			Operator: corev1.TolerationOpEqual,
+			Value:    "taint-value",
+			Effect:   corev1.TaintEffectNoSchedule,
+		},
+	}
+
+	var testCases = []struct {
+		testName             string
+		helmValues           map[string]string
+		expectedNodeSelector map[string]string
+		expectedTolerations  []corev1.Toleration
+	}{
+		{
+			testName: "Both nodeSelector and tolerations are present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"nodeSelector.selector-key":   "selector-value",
+				"tolerations[0].key":          "taint-key",
+				"tolerations[0].operator":     "Equal",
+				"tolerations[0].value":        "taint-value",
+				"tolerations[0].effect":       "NoSchedule",
+			},
+			expectedNodeSelector: nodeSelector,
+			expectedTolerations:  toleration,
+		},
+		{
+			testName: "Only nodeSelector is present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"nodeSelector.selector-key":   "selector-value",
+			},
+			expectedNodeSelector: nodeSelector,
+			expectedTolerations:  []corev1.Toleration{},
+		},
+		{
+			testName: "Only tolerations is present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"tolerations[0].key":          "taint-key",
+				"tolerations[0].operator":     "Equal",
+				"tolerations[0].value":        "taint-value",
+				"tolerations[0].effect":       "NoSchedule",
+			},
+			expectedNodeSelector: map[string]string{},
+			expectedTolerations:  toleration,
+		},
+		{
+			testName: "Both nodeSelector and tolerations are not present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+			},
+			expectedNodeSelector: map[string]string{},
+			expectedTolerations:  []corev1.Toleration{},
+		},
+	}
+	for _, tc := range testCases {
+		c.Logf("Test name:%s ", tc.testName)
+		defer func() {
+			h.helmApp.dryRun = false
+		}()
+		// Installing kanister release from local kanister-operator - Dry run"
+		testApp, err := NewHelmApp(tc.helmValues, kanisterName, "../../../helm/kanister-operator", kanisterName, "", true)
+		c.Assert(err, IsNil)
+
+		out, err := testApp.Install()
+		c.Assert(err, IsNil)
+		components := helm.ResourcesFromRenderedManifest(out, func(kind helm.K8sObjectType) bool {
+			return kind == helm.K8sObjectTypeDeployment
+		})
+		c.Assert(len(components) > 0, Equals, true)
+		// Take the deployment components
+		mapOfObjs, err := helm.GetK8sObjectsFromRenderedManifest[*appsv1.Deployment](components)
+		c.Assert(err, IsNil)
+		// Use only the required deployment
+		var obj = mapOfObjs[h.deploymentName]
+		c.Assert(obj, NotNil)
+
+		c.Assert(len(obj.Spec.Template.Spec.NodeSelector), Equals, len(tc.expectedNodeSelector))
+		for k, v := range tc.expectedNodeSelector {
+			c.Assert(obj.Spec.Template.Spec.NodeSelector[k], Equals, v)
+		}
```

Comment:
can we not just match the nodeSelector of deployment and the nodeselector of test directly instead of matching the keys separately?

Reply:
AFAIU these being reference types, there is no way of comparing them for equality. Maps are considered equal if and only if they contain the same keys with the same corresponding values.